### ISSUE TITLE
Add idle timeout for upstream connections

### ISSUE THREAD

**Issue by JonathanO** [envoyproxy/envoy#2324]
If a request is sent to an upstream just after the Keepalive timeout has expired Envoy will detect this as a reset connection. #1979 suggested adding a retry to handle this case, however there is no safe way to retry reset connections, since we have no indication of whether or not the upstream started processing the (possibly non-idempotent) request.

To avoid this scenario, we should add the ability to configure an idle timeout for upstream connections. If this is configured to be less than the upstream's keep alive timeout (taking max RTT into account!) then you should never see upstream resets due to keepalive timeouts under normal circumstances.

I propose adding connection_idle_timeout to cluster configuration (alongside the existing max_requests_per_connection.) Idle time should be measured from the last byte read from the connection. A connection that has exceeded the idle timeout should not be used for any new requests. There'll presumably also need to be some mechanism that closes and cleans up timed out connections.

**Comment by baranov1ch** [comment]
>since we have no indication of whether or not the upstream started processing the (possibly non-idempotent) request

BTW, we can check if anything was written to the connection, and safely retry these cases (this is what golang net/http does)

**Comment by JonathanO** [comment]
That's, very unfortunately, not correct (and also not what golang net/http does.) Receiving a FIN from the server will cause reads on the client side to return EOF, but writes will continue to succeed. Some clients attempt to deal with this by performing a non-blocking read to check for EOF before reusing a connection, however that still leaves a significant window where the client's test read can succeed, but the server has closed the connection by the time the request being sent arrives there.

As for golang net/http the relevant code to decide whether or not to retry a failed request is here: https://github.com/golang/go/blob/master/src/net/http/transport.go#L454
It handles the described scenario on line 492, by retrying if the attempt to READ the response returned EOF immediately, however it only does this if req.isReplayable() (line 483), which is true only for 
idempotent requests (https://github.com/golang/go/blob/master/src/net/http/request.go#L1308)
In short, a server keep-alive timeout with net/http will still lead to non-idempotent requests failing in this way.

**Comment by mattklein123** [comment]
Yup closing.
